src/Microsoft.Health.TaskManagement/IQueueClient.cs

```diff
@@ -106,5 +99,28 @@ public interface IQueueClient
         /// <param name="requestCancellationOnFailure">Cancel other jobs with same group id if this job failed.</param>
         /// <param name="cancellationToken">Cancellation token</param>
         public Task CompleteJobAsync(JobInfo jobInfo, bool requestCancellationOnFailure, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Executes job with heartbeats. Put heartbeat logic saves Data and Result from provided jobInfo.
+        /// Usage of this method should be discouraged. Please use method with lightweight heartbeats.
+        /// </summary>
+        /// <param name="jobInfo">job info</param>
+        /// <param name="action">job action</param>
+        /// <param name="heartbeatPeriod">heartbeat period in seconds</param>
+        /// <param name="cancellationTokenSource">cancellationTokenSource</param>
+        /// <returns>result</returns>
+        public Task<string> ExecuteJobWithHeartbeats(JobInfo jobInfo, Func<CancellationTokenSource, Task<string>> action, TimeSpan heartbeatPeriod, CancellationTokenSource cancellationTokenSource);
```

Comment:
I don't think it is appropriate to put keep heartbeat timer in queue client. Queue client should be only just storage level implementation. let's put the timer in the job hosting.